### ISSUE TITLE
每个技巧的影响(跨域)

### ISSUE THREAD

**Issue by 2110317008** [michuanhaohao/reid-strong-baseline#76]
从M->D和从D_>M，就是如何在一个数据集上训练在另一个数据集上测试？

**Comment by shaoniangu** [comment]
Hi，测试的时候确定好要加载的数据集和使用的训练好的ckpt就可以了呀。

**Comment by 2110317008** [comment]
> Hi，测试的时候确定好要加载的数据集和使用的训练好的ckpt就可以了呀。
您好！
ckpt已经在训练集上训好，当我们运行所提提供的那条命令，如：python3 tools/train.py --config_file='configs/softmax_triplet.yml' MODEL.DEVICE_ID "('your device id')" DATASETS.NAMES "('market1501')" OUTPUT_DIR "('your path to save checkpoints and logs')"   当训完120epoch后就紧接着就是提取特征、测试，测试数据集的加载，完事后就是mAP和rank。
请问您尝试过吗？

**Comment by 2110317008** [comment]
> Hi，测试的时候确定好要加载的数据集和使用的训练好的ckpt就可以了呀。

ckpt已经在训练集上训好，当我们运行所提提供的那条命令，如：python3 tools/train.py --config_file='configs/softmax_triplet.yml' MODEL.DEVICE_ID "('your device id')" DATASETS.NAMES "('market1501')" OUTPUT_DIR "('your path to save checkpoints and logs')" 当训完120epoch后就紧接着就是提取特征、测试，测试数据集的加载，完事后就是mAP和rank。
请问您尝试过一个数据集上训练在另一个数据集上测试吗？

**Comment by shaoniangu** [comment]
尝试过的，论文里给出了cross domain的性能的。你测试的话应该运行test脚本的呀。想要训完120epoch直接做跨域测试，你需要修改evaluator的dataloader为目标测试集的dataloader。

**Comment by 2110317008** [comment]
非常感谢！明白了！

**Comment by shaoniangu** [comment]
应该的~

**Comment by 2110317008** [comment]
看完代码我发现
在./data/build.py下面的make_data_load(cfg)下加载了
dataset = init_dataset(cfg.DATASETS.NAMES, root=cfg.DATASETS.ROOT_DIR)
train_set = ImageDataset(dataset.train, train_transforms)
train_loader=DataLoader()
到这里我认为训练集就加载完毕了
接下来
val_set = ImageDataset(dataset.query + dataset.gallery, val_transforms)  #到里我认为仍然加载的是所训练数据集下面的测试集
val_loader = DataLoader()
到这里我认为测试数据集也加载完毕了

我想改的就是在加载测试集之前把dataset=另一个数据集，我就把_C.DATASETS.NAMES=('market1501')下面又加了一个_C.DATASETS_t.NAMES = ('dukemtmc')，然后就在加载测试集之前把dataset=init_dataset(cfg.DATASETS_t.NAMES, root=cfg.DATASETS.ROOT_DIR)   

但是在我运行CUDA_VISIBLE_DEVICES=0 python3 tools/train.py --config_file='configs/softmax_triplet.yml' MODEL.DEVICE_ID "('your device id')" DATASETS.NAMES "('market1501')" OUTPUT_DIR "('your path to save checkpoints and logs')"
报错
初学者，希望指教/微笑/
@shaoniangu

**Comment by shaoniangu** [comment]
我感觉应该是没有问题的，应该是替换了val_loader就好，因为测试脚本和训练时候的验证脚本是差不多的，请仔细对比检查一下，确认修改的部分返回值以及create_evaluator的形参有没有问题，如果没有的话，请给我看一下报错的log。

**Comment by 2110317008** [comment]
def make_data_loader(cfg):
    train_transforms = build_transforms(cfg, is_train=True)
    val_transforms = build_transforms(cfg, is_train=False)
    num_workers = cfg.DATALOADER.NUM_WORKERS# Number of data loading threads  _C.DATALOADER.NUM_WORKERS = 8   _C as cfg
    if len(cfg.DATASETS.NAMES) == 1:
        dataset = init_dataset(cfg.DATASETS.NAMES, root=cfg.DATASETS.ROOT_DIR)
    else:
        # TODO: add multi dataset to train
        dataset = init_dataset(cfg.DATASETS.NAMES, root=cfg.DATASETS.ROOT_DIR)

    num_classes = dataset.num_train_pids
    train_set = ImageDataset(dataset.train, train_transforms)
    if cfg.DATALOADER.SAMPLER == 'softmax':
        train_loader = DataLoader(
            train_set, batch_size=cfg.SOLVER.IMS_PER_BATCH, shuffle=True, num_workers=num_workers,
            collate_fn=train_collate_fn
        )
    else:
        train_loader = DataLoader(
            train_set, 
            batch_size=cfg.SOLVER.IMS_PER_BATCH,
            sampler=RandomIdentitySampler(dataset.train, cfg.SOLVER.IMS_PER_BATCH, cfg.DATALOADER.NUM_INSTANCE),
            # sampler=RandomIdentitySampler_alignedreid(dataset.train, cfg.DATALOADER.NUM_INSTANCE),      # new add by gu
            num_workers=num_workers, collate_fn=train_collate_fn
        )
    if cfg.DATASETS.NAMES == 'market1501':
        DATASETS_t_NAMES = 'dukemtmc'
    if cfg.DATASETS.NAMES == 'dukemtmc':
        DATASETS_t_NAMES = 'market1501'
    dataset_temp = init_dataset(DATASETS_t_NAMES, root=cfg.DATASETS.ROOT_DIR)
    val_set = ImageDataset(dataset_temp.query + dataset_temp.gallery, val_transforms) 
    val_loader = DataLoader(
        val_set, 
        batch_size=cfg.TEST.IMS_PER_BATCH, 
        shuffle=False, 
        num_workers=num_workers,
        collate_fn=val_collate_fn
    )
    return train_loader, val_loader, len(dataset.query), num_classes


 val_loader = DataLoader()这个里面的第一个参数val_set就是代表了测试的数据集，，
 val_set = ImageDataset()里面加载dataset_temp的query和gallery,
上面那个报错是我在default.py文件里面加了个_C.DATASETS_t.NAMES = ('dukemtmc')，然后报出找不到_C.DATASETS_t.NAMES，

这次我这样该没出错，也好像加载了两个个数据集但是结果差的多，下面是输出命令后出现的
TEST:
  FEAT_NORM: yes
  IMS_PER_BATCH: 128
  NECK_FEAT: after
  RE_RANKING: no
  WEIGHT: path
This dataset has been downloaded.
=> DukeMTMC-reID loaded
Dataset statistics:
  ----------------------------------------
  subset   | # ids | # images | # cameras
  ----------------------------------------
  train    |   702 |    16522 |         8
  query    |   702 |     2228 |         8
  gallery  |  1110 |    17661 |         8
  ----------------------------------------
=> Market1501 loaded
Dataset statistics:
  ----------------------------------------
  subset   | # ids | # images | # cameras
  ----------------------------------------
  train    |   751 |    12936 |         6
  query    |   750 |     3368 |         6
  gallery  |   751 |    15913 |         6
  ----------------------------------------
Loading pretrained ImageNet model......
Train with center loss, the loss type is triplet_center
label smooth on, numclasses: 702

**Comment by shaoniangu** [comment]
看起来这样貌似没啥问题呀，跨域的点确实是很低的，你可以对比论文看看。

**Comment by 2110317008** [comment]
> 看起来这样貌似没啥问题呀，跨域的点确实是很低的，你可以对比论文看看。

五点之内都能接受，but 我这样得出的结果和论文中的数据相差甚多

**Comment by michuanhaohao** [comment]
@2110317008 能把结果放出来吗？我貌似没找到你的结果

**Comment by 2110317008** [comment]
> @2110317008 能把结果放出来吗？我貌似没找到你的结果

后面我又跑了2遍有点接近了，就没管了。看来以后再做实验跑程序的时候得做记录了！

**Comment by michuanhaohao** [comment]
好的，没问题就好
